signac/contrib/migration/__init__.py

```diff
@@ -5,102 +5,141 @@
-    project_reloaded = project.get_project(
-        root=project.root_directory(), search=False, _ignore_schema_version=True
-    )
-    project._config = project_reloaded._config
+_MIGRATIONS = {
+    ("0", "1"): _migrate_v0_to_v1,
+}
 
 
-def _update_project_config(project, **kwargs):
-    """Update the project configuration."""
-    for fn in ("signac.rc", ".signacrc"):
-        config = get_config(project.fn(fn))
-        if "project" in config:
-            break
-    else:
-        raise RuntimeError("Unable to determine project configuration file.")
-    config.update(kwargs)
-    config.write()
-    _reload_project_config(project)
+_PARSED_SCHEMA_VERSION = version.parse(SCHEMA_VERSION)
 
+_PARSED_VERSION_LIST = list(
+    reversed(sorted(version.parse(v) for v in _CONFIG_LOADERS.keys()))
+)
 
-@contextmanager
-def _lock_for_migration(project):
-    lock = FileLock(project.fn(FN_MIGRATION_LOCKFILE))
-    try:
-        with lock:
-            yield
-    finally:
+
+def _get_config_schema_version(root_directory, version_guess):
+    # Try loading the schema using the loader corresponding to the expected
+    # version if it has a configured loader.
+    versions = _PARSED_VERSION_LIST
+    if version_guess in _CONFIG_LOADERS:
+        versions = [version_guess] + versions
+    for guess in versions:
         try:
-            os.unlink(lock.lock_file)
-        except FileNotFoundError:
+            # Note: We could consider using a different component as the key
+            # for _CONFIG_LOADERS, but since this is an internal detail it's
+            # not terribly consequential.
+            config = _CONFIG_LOADERS[guess.public](root_directory)
+            break
+        except Exception:
+            # The load failed, go to the next
             pass
+    else:
+        raise RuntimeError("Unable to load config file.")
+    try:
+        return version.parse(config["schema_version"])
+    except KeyError:
+        # The default schema version is version 0.
+        return version.parse("0")
 
 
-def _collect_migrations(project):
-    schema_version = version.parse(SCHEMA_VERSION)
-
-    def get_config_schema_version():
-        return version.parse(project._config["schema_version"])
+def _collect_migrations(root_directory):
+    schema_version = _PARSED_SCHEMA_VERSION
 
-    if get_config_schema_version() > schema_version:
+    current_schema_version = _get_config_schema_version(
+        root_directory, _PARSED_SCHEMA_VERSION
+    )
+    if current_schema_version > schema_version:
         # Project config schema version is newer and therefore not supported.
         raise RuntimeError(
-            "The signac schema version used by this project is {}, but signac {} "
-            "only supports up to schema version {}. Try updating signac.".format(
-                get_config_schema_version(), __version__, SCHEMA_VERSION
-            )
+            f"The signac schema version used by this project is "
```

Comment:
This might be considered lint by some tools. It's a f-string with no fields.
```suggestion
            "The signac schema version used by this project is "
```